okio-files/src/commonMain/kotlin/okio/Path.kt

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio
+
+import okio.ByteString.Companion.EMPTY
+import okio.ByteString.Companion.encodeUtf8
+
+/**
+ * A hierarchical address on a filesystem. A path is an identifier only; a [Filesystem] is required
+ * to access the file that a path refers to, if any.
+ *
+ * Paths may be absolute or relative:
+ *
+ * * **Absolute paths** are prefixed with `/` and identify a location independent of any working
+ *   directory.
+ * * **Relative paths** are not prefixed with `/`. On their own, relative paths do not identify a
+ *   location on a filesystem; they must be resolved against an absolute path first. When a relative
+ *   path is used to access a [Filesystem], it is resolved against [Filesystem.cwd].
+ *
+ * After the optional leading `/`, the rest of the path is a sequence of segments separated by `/`
+ * characters. Segments satisfy these rules:
+ *
+ *  * Segments are always non-empty.
+ *  * Segments may not be `.`.
+ *  * Segments with the value `..` only occur in relative paths, where these must precede all other
+ *    segments.
+ *
+ * The only path that ends with `/` is the filesystem root, `/`. The empty path `""` is a relative
+ * path that resolves to whichever path it is resolved against.
+ *
+ * Sample Paths
+ * ------------
+ *
+ * | Path                   | Type       | Parent              |
+ * | :--------------------- | :--------- | :------------------ |
+ * | `/Users/jessewilson`   | Absolute   | `/Users`            |
+ * | `/Users`               | Absolute   | `/`                 |
+ * | `/`                    | Absolute   | null                |
+ * | `src/main/java`        | Relative   | `src/main`          |
+ * | `src/main`             | Relative   | `src`               |
+ * | `src`                  | Relative   | (empty)             |
```

Comment:
Why is the parent of a relative path empty and not null?

Reply:
Probably best is to pattern match.

```
mkdir -p ~/xxx/yyy/zzz

cd ~/
cd "xxx/yyy/zzz"
pwd
// prints ~/xxx/yyy/zzz

cd ~/
cd "xxx/yyy"
pwd
// prints ~/xxx/yyy

cd ~/
cd "xxx"
pwd
// prints ~/xxx

cd ~/
cd ""
pwd
// prints ~/
```

The empty relative resolves to whatever it was resolved against.